File: src/cutting/constructors.py
"""Candidate bin-fill generators for the board-count search.

Each generator packs ONE bin from a pool of remaining pieces and returns a
``BinFill`` (placements + leftovers + cuts + a dedupe signature). The search
layer (``search.py``) evaluates many candidates per board and keeps the ones
that lead to the cheapest global solution — no single generator is trusted:
the audit showed every greedy fails at its own scale.

Two families:

- ``greedy_fill``: the classic one-pass packer under a ``GreedyConfig``
  (piece order x free-rect selection x split rule). Cheap and strong on
  homogeneous jobs.
- ``strip_fill``: builds the bin as first-stage guillotine strips (the
  structure commercial optimizers emit): candidate strip sizes come from the
  remaining piece dimensions plus the bin's residual width, and each strip is
  packed recursively with the guillotine packer, which yields 2/3-stage
  patterns (sub-columns inside strips). ``first_dim`` seeds diversity: forcing
  the first strip lets the search explore layouts a pure efficiency-greedy
  would never open with (e.g. distributing full-length thin strips across
  boards instead of hoarding them into one).
"""

from dataclasses import dataclass
from typing import Callable, Dict, List, Optional, Sequence, Tuple

from src.cutting.enums import PackingStrategy, SplitRule
from src.cutting.models import BinSpec, Cut, Material, Piece, PlacedPiece, Rectangle
from src.cutting.packer import GuillotineOptimizer
from src.cutting.parameters import CuttingParameters

# Piece-type identity: geometry + constraints. Two pieces of the same type are
# interchangeable for the search (ids only matter for downstream attribution).
PieceType = Tuple[float, float, bool, int]


def piece_type(piece: Piece) -> PieceType:
    return (piece.width, piece.height, piece.can_rotate, piece.priority)


SORT_KEYS: Dict[str, Callable[[Piece], tuple]] = {
    "area": lambda p: (-p.priority, -p.area, -p.height, p.id),
    "maxdim": lambda p: (
        -p.priority,
        -max(p.width, p.height),
        -min(p.width, p.height),
        p.id,
    ),
    "height": lambda p: (-p.priority, -p.height, -p.width, p.id),
    "width": lambda p: (-p.priority, -p.width, -p.height, p.id),
    "perimeter": lambda p: (-p.priority, -(p.width + p.height), -p.area, p.id),
    "mindim": lambda p: (
        -p.priority,
        -min(p.width, p.height),
        -max(p.width, p.height),
        p.id,
    ),
}


@dataclass(frozen=True)
class GreedyConfig:
    """One point of the greedy portfolio: order x selection x split."""

    sort: str
    split: SplitRule
    # ``MAX_EFFICIENCY`` selects free rects by Best-Area-Fit, ``LONG_OFFCUTS``
    # by Bottom-Left; reused here as the selection axis of the portfolio.
    selection: PackingStrategy = PackingStrategy.MAX_EFFICIENCY


# Deterministic portfolio order. Curated: the full cartesian product mostly
# yields duplicate fills (dedupe absorbs them) but costs decodes; this set
# covers the distinct behaviors observed in the audit experiments.
GREEDY_PORTFOLIO: Tuple[GreedyConfig, ...] = tuple(
    GreedyConfig(sort=s, split=sp, selection=sel)
    for s in ("area", "maxdim", "height", "width", "perimeter", "mindim")
    for sp in (
        SplitRule.SHORTER_LEFTOVER_AXIS,
        SplitRule.LONGER_AXIS,
        SplitRule.MINIMIZE_AREA,
        SplitRule.MAXIMIZE_AREA,
    )
    for sel in (PackingStrategy.MAX_EFFICIENCY, PackingStrategy.LONG_OFFCUTS)
)


@dataclass
class BinFill:
    """One candidate fill of one bin."""

    spec: BinSpec
    placed: List[PlacedPiece]
    remainders: List[Rectangle]
    cuts: List[Cut]

    @property
    def used_area(self) -> float:
        return sum(p.width * p.height for p in self.placed)

    @property
    def waste_area(self) -> float:
        return self.spec.area - self.used_area

    def type_signature(self) -> Tuple[Tuple[PieceType, int], ...]:
        """Multiset of piece types placed — the dedupe identity of a fill."""
        counts: Dict[PieceType, int] = {}
        for pp in self.placed:
            t = piece_type(pp.piece)
            counts[t] = counts.get(t, 0) + 1
        return tuple(sorted(counts.items()))

    def placed_ids(self) -> List[str]:
        return [pp.piece.id for pp in self.placed]


def greedy_fill(
    pool: Sequence[Piece],
    spec: BinSpec,
    cutting_params: CuttingParameters,
    config: GreedyConfig,
    min_rect_size: float = 0.1,
) -> Optional[BinFill]:
    """One-pass greedy fill of ``spec`` under ``config``; ``None`` if nothing fits."""
    try:
        packer = GuillotineOptimizer(
            material=spec.to_material(),
            cutting_params=cutting_params,
            split_rule=config.split,
            strategy=config.selection,
            min_rect_size=min_rect_size,
        )
    except ValueError:
        # Trims exceed the bin dimensions (tiny offcut): unusable.
        return None
    ordered = sorted(pool, key=SORT_KEYS[config.sort])
    placed, _ = packer.optimize(ordered, presorted=True)
    if not placed:
        return None
    return BinFill(
        spec=spec,
        placed=placed,
        remainders=packer.remainders,
        cuts=packer.cuts,
    )


def _strip_candidate_dims(
    pool: Sequence[Piece], limit: float, across: bool
) -> List[float]:
    """Distinct piece dimensions usable as strip sizes, descending.

    ``across=False`` collects the dimension perpendicular to the strip's long
    axis for vertical strips (piece widths, plus heights of rotatable pieces);
    ``across=True`` the same for horizontal strips.
    """
    dims = set()
    for p in pool:
        primary = p.height if across else p.width
        secondary = p.width if across else p.height
        if primary <= limit:
            dims.add(primary)
        if p.can_rotate and secondary <= limit:
            dims.add(secondary)
    return sorted(dims, reverse=True)


def _strip_material(spec: BinSpec, width: float, height: float) -> Material:
    """Strip-sized pseudo material for the nested packer (id irrelevant)."""
    return Material(
        id=spec.key,
        width=width,
        height=height,
        thickness=spec.thickness,
        cost_per_unit=0.0,
    )


def _translate(
    placed: List[PlacedPiece],
    cuts: List[Cut],
    remainders: List[Rectangle],
    dx: float,
    dy: float,
) -> Tuple[List[PlacedPiece], List[Cut], List[Rectangle]]:
    moved_pieces = [
        PlacedPiece(
            piece=pp.piece,
            x=pp.x + dx,
            y=pp.y + dy,
            width=pp.width,
            height=pp.height,
            rotated=pp.rotated,
        )
        for pp in placed
    ]
    moved_cuts = [
        Cut(c.x + dx, c.y + dy, c.length, is_horizontal=c.is_horizontal) for c in cuts
    ]
    moved_rects = [Rectangle(r.x + dx, r.y + dy, r.width, r.height) for r in remainders]
    return moved_pieces, moved_cuts, moved_rects


def strip_fill(
    pool: Sequence[Piece],
    spec: BinSpec,
    cutting_params: CuttingParameters,
    horizontal: bool = False,
    first_dim: Optional[float] = None,
    max_repeat: Optional[int] = None,
    min_rect_size: float = 0.1,
) -> Optional[BinFill]:
    """Fills ``spec`` as a sequence of first-stage guillotine strips.

    Vertical strips span the usable height and consume width (a table-saw rip
    along the long axis); ``horizontal=True`` mirrors the roles (the first
    stage is a crosscut). Each strip is a nested recursive guillotine pack, so
    strips naturally contain sub-columns (3-stage patterns). Strip sizes are
    chosen by packed efficiency among the remaining piece dimensions plus the
    bin residual, which lets narrow leftovers host paired sub-columns.

    ``max_repeat`` caps how many strips with an identical piece-type content
    one board commits. A perfect single-type strip (e.g. a full-length 70mm
    rail, 100% efficient) greedily hoards every sibling into one board — the
    exact move that strands the rest of the job; capping repeats yields the
    candidates that SPREAD those strips across boards, and the search decides
    which distribution wins.
    """
    params = cutting_params or CuttingParameters()
    kerf = max(0, params.kerf)
    left, right = max(0, params.left_trim), max(0, params.right_trim)
    top, bottom = max(0, params.top_trim), max(0, params.bottom_trim)
    usable_w = spec.width - left - right
    usable_h = spec.height - top - bottom
    if usable_w <= 0 or usable_h <= 0:
        return None

    # Work in strip-local coordinates: strips advance along ``axis_room``.
    axis_room = usable_h if horizontal else usable_w
    strip_span = usable_w if horizontal else usable_h

    remaining = list(pool)
    all_placed: List[PlacedPiece] = []
    all_cuts: List[Cut] = []
    all_rects: List[Rectangle] = []
    cursor = 0.0
    room = axis_room
    first = True
    repeat_counts: Dict[Tuple[Tuple[PieceType, int], ...], int] = {}

    strip_params = CuttingParameters(kerf=kerf)

    while remaining and room >= min_rect_size:
        dims = _strip_candidate_dims(remaining, room, across=horizontal)
        if not dims:
            break
        candidates = list(dims)
        if room not in candidates:
            candidates.append(room)
        if first and first_dim is not None:
            if first_dim <= room:
                candidates = [first_dim]
            first = False

        best = None
        for size in candidates:
            if horizontal:
                mat_w, mat_h = strip_span, size
            else:
                mat_w, mat_h = size, strip_span
            try:
                packer = GuillotineOptimizer(
                    material=_strip_material(spec, mat_w, mat_h),
                    cutting_params=strip_params,
                    min_rect_size=min_rect_size,
                )
            except ValueError:
                continue
            fitting = [
                p
                for p in remaining
                if (p.width <= mat_w and p.height <= mat_h)
                or (p.can_rotate and p.height <= mat_w and p.width <= mat_h)
            ]
            if not fitting:
                continue

            # Widest-fit-first inside the strip (rotation-aware) wastes the
            # least strip size; the nested packer still nests narrower
            # sub-columns recursively.
            def _fit_across(p: Piece, limit: float, across_height: bool) -> float:
                natural = p.height if across_height else p.width
                rotated = p.width if across_height else p.height
                dims = []
                if natural <= limit:
                    dims.append(natural)
                if p.can_rotate and rotated <= limit:
                    dims.append(rotated)
                return max(dims) if dims else 0.0

            if horizontal:
                fitting.sort(
                    key=lambda p: (
                        -p.priority,
                        -_fit_across(p, mat_h, True),
                        -p.width,
                        p.id,
                    )
                )
            else:
                fitting.sort(
                    key=lambda p: (
                        -p.priority,
                        -_fit_across(p, mat_w, False),
                        -p.height,
                        p.id,
                    )
                )
            placed, _ = packer.optimize(fitting, presorted=True)
            if not placed:
                continue
            if max_repeat is not None:
                sig_counts: Dict[PieceType, int] = {}
                for pp in placed:
                    t = piece_type(pp.piece)
                    sig_counts[t] = sig_counts.get(t, 0) + 1
                strip_sig = tuple(sorted(sig_counts.items()))
                if repeat_counts.get(strip_sig, 0) >= max_repeat:
                    continue
            else:
                strip_sig = None
            used = sum(pp.width * pp.height for pp in placed)
            eff = used / (mat_w * mat_h)
            score = (eff, used, -size)
            if best is None or score > best[0]:
                best = (score, size, placed, packer, strip_sig)
        if best is None:
            break

        _, size, placed, packer, strip_sig = best
        if strip_sig is not None:
            repeat_counts[strip_sig] = repeat_counts.get(strip_sig, 0) + 1
        if horizontal:
            dx, dy = left, bottom + cursor
        else:
            dx, dy = left + cursor, bottom
        moved_pieces, moved_cuts, moved_rects = _translate(
            placed, packer.cuts, packer.remainders, dx, dy
        )
        all_placed.extend(moved_pieces)
        all_cuts.extend(moved_cuts)
        all_rects.extend(moved_rects)

        leftover = room - size
        if leftover > 0:
            # First-stage cut separating this strip from the rest of the bin.
            if horizontal:
                all_cuts.append(
                    Cut(left, bottom + cursor + size, usable_w, is_horizontal=True)
                )
            else:
                all_cuts.append(
                    Cut(left + cursor + size, bottom, usable_h, is_horizontal=False)
                )
            advance = size + kerf
        else:
            advance = size
        cursor += advance
        room -= advance

        placed_ids = {pp.piece.id for pp in placed}
        remaining = [p for p in remaining if p.id not in placed_ids]

    if not all_placed:
        return None

    if room >= min_rect_size:
        # Unopened tail of the bin: one continuous first-stage leftover.
        if horizontal:
            all_rects.append(Rectangle(left, bottom + cursor, usable_w, room))
        else:
            all_rects.append(Rectangle(left + cursor, bottom, room, usable_h))

    return BinFill(spec=spec, placed=all_placed, remainders=all_rects, cuts=all_cuts)

File: rust/src/constructors.rs
//! Port of the parts of `src/cutting/constructors.py` the packer needs:
//! the six sort comparators, `greedy_fill` and `strip_fill`.

use std::collections::{HashMap, HashSet};

use crate::models::{Cut, CuttingParams, Piece, PlacedPiece, Rect, Selection, SplitRule};
use crate::packer::Packer;

/// Codes for `constructors.SORT_KEYS`, in the order `GREEDY_PORTFOLIO`
/// iterates them: area, maxdim, height, width, perimeter, mindim.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SortKey {
    Area = 0,
    MaxDim = 1,
    Height = 2,
    Width = 3,
    Perimeter = 4,
    MinDim = 5,
}

impl SortKey {
    pub fn from_code(code: u8) -> Option<Self> {
        match code {
            0 => Some(SortKey::Area),
            1 => Some(SortKey::MaxDim),
            2 => Some(SortKey::Height),
            3 => Some(SortKey::Width),
            4 => Some(SortKey::Perimeter),
            5 => Some(SortKey::MinDim),
            _ => None,
        }
    }

    /// The two float slots of the comparator, after `-priority` and before the
    /// id tiebreak. Every key in `SORT_KEYS` has exactly this shape.
    #[inline]
    fn dims(&self, p: &Piece) -> (f64, f64) {
        let (w, h) = (p.width, p.height);
        let (lo, hi) = if w < h { (w, h) } else { (h, w) };
        match self {
            SortKey::Area => (-p.area, -h),
            SortKey::MaxDim => (-hi, -lo),
            SortKey::Height => (-h, -w),
            SortKey::Width => (-w, -h),
            SortKey::Perimeter => (-(w + h), -p.area),
            SortKey::MinDim => (-lo, -hi),
        }
    }
}

/// `sorted(pool, key=SORT_KEYS[...])`.
///
/// The comparator is total — every key ends in the piece id — so stability is
/// not load-bearing here, but `sort_by` is stable anyway and that matches
/// Python's `sorted`. `id_rank` stands in for the id string: it is the rank of
/// that id in the pool's lexicographic order, so comparing ranks reproduces
/// Python's string comparison without moving strings across the boundary.
pub fn sort_pool(pieces: &mut [Piece], key: SortKey) {
    pieces.sort_by(|a, b| {
        let pa = -(a.priority as f64);
        let pb = -(b.priority as f64);
        pa.partial_cmp(&pb)
            .unwrap()
            .then_with(|| {
                let (a1, a2) = key.dims(a);
                let (b1, b2) = key.dims(b);
                a1.partial_cmp(&b1)
                    .unwrap()
                    .then_with(|| a2.partial_cmp(&b2).unwrap())
            })
            .then_with(|| a.id_rank.cmp(&b.id_rank))
    });
}

pub struct Fill {
    pub placed: Vec<crate::models::PlacedPiece>,
    pub remainders: Vec<crate::models::Rect>,
    pub cuts: Vec<crate::models::Cut>,
}

/// Port of `constructors.greedy_fill`. `None` mirrors both of its exits: the
/// trims-exceed-the-bin `ValueError`, and "nothing fit".
pub fn greedy_fill(
    pool: &[Piece],
    spec_width: f64,
    spec_height: f64,
    params: &CuttingParams,
    sort_key: SortKey,
    split_rule: SplitRule,
    selection: Selection,
    min_rect_size: f64,
) -> Option<Fill> {
    let mut packer = Packer::new(
        spec_width,
        spec_height,
        params,
        split_rule,
        selection,
        min_rect_size,
    )?;

    let mut ordered: Vec<Piece> = pool.to_vec();
    sort_pool(&mut ordered, sort_key);
    packer.optimize(&ordered);

    if packer.placed.is_empty() {
        return None;
    }
    Some(Fill {
        placed: packer.placed,
        remainders: packer.remainders,
        cuts: packer.cuts,
    })
}

// ---------------------------------------------------------------------------
// strip_fill
// ---------------------------------------------------------------------------

/// Mirror of `constructors.piece_type` — geometry + constraints, hashable.
///
/// The dimensions are held as bit patterns because `f64` is not `Hash`. Every
/// dimension is strictly positive, so bit order *is* value order, and the
/// derived `Ord` therefore sorts a signature exactly like Python's
/// `sorted(counts.items())` — though only canonicality matters, since the
/// signature is used solely as the `max_repeat` bookkeeping key.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
struct PieceType {
    width: u64,
    height: u64,
    can_rotate: bool,
    priority: i32,
}

type TypeSignature = Vec<(PieceType, u32)>;

#[inline]
fn piece_type(p: &Piece) -> PieceType {
    PieceType {
        width: p.width.to_bits(),
        height: p.height.to_bits(),
        can_rotate: p.can_rotate,
        priority: p.priority,
    }
}

/// Multiset of the *original* piece types behind a set of placements.
///
/// `PlacedPiece` carries the rotated dimensions, so the type has to be read off
/// the pool entry (`pool[index]`), exactly as Python reads `pp.piece`.
fn type_signature(placed: &[PlacedPiece], pool: &[Piece]) -> TypeSignature {
    let mut counts: HashMap<PieceType, u32> = HashMap::new();
    for pp in placed {
        *counts.entry(piece_type(&pool[pp.index as usize])).or_insert(0) += 1;
    }
    let mut sig: TypeSignature = counts.into_iter().collect();
    sig.sort_unstable();
    sig
}

/// Port of `_strip_candidate_dims`: distinct usable strip sizes, descending.
///
/// `sort` + `dedup` reproduces `sorted(set(dims), reverse=True)`.
fn strip_candidate_dims(pool: &[Piece], limit: f64, across: bool) -> Vec<f64> {
    let mut dims: Vec<f64> = Vec::with_capacity(pool.len() * 2);
    for p in pool {
        let (primary, secondary) = if across {
            (p.height, p.width)
        } else {
            (p.width, p.height)
        };
        if primary <= limit {
            dims.push(primary);
        }
        if p.can_rotate && secondary <= limit {
            dims.push(secondary);
        }
    }
    dims.sort_by(|a, b| b.partial_cmp(a).unwrap());
    dims.dedup();
    dims
}

/// Port of the nested `_fit_across`: the widest orientation that still fits
/// across the strip, or 0.0 when neither does.
#[inline]
fn fit_across(p: &Piece, limit: f64, across_height: bool) -> f64 {
    let (natural, rotated) = if across_height {
        (p.height, p.width)
    } else {
        (p.width, p.height)
    };
    let mut best = 0.0_f64;
    if natural <= limit {
        best = natural;
    }
    if p.can_rotate && rotated <= limit && rotated > best {
        best = rotated;
    }
    best
}

/// `(eff, used, -size) > best`, the strip's selection comparison. Strict, so
/// the FIRST candidate wins an exact tie — as in Python.
#[inline]
fn score_gt(a: (f64, f64, f64), b: (f64, f64, f64)) -> bool {
    if a.0 != b.0 {
        return a.0 > b.0;
    }
    if a.1 != b.1 {
        return a.1 > b.1;
    }
    a.2 > b.2
}

/// One strip kept as the round's winner: everything Python reads back off its
/// `best` tuple (the packer included, via its three output lists).
struct Strip {
    score: (f64, f64, f64),
    size: f64,
    placed: Vec<PlacedPiece>,
    cuts: Vec<Cut>,
    remainders: Vec<Rect>,
    signature: Option<TypeSignature>,
}

/// Port of `constructors.strip_fill`.
///
/// Fills the bin as a sequence of first-stage guillotine strips, each packed
/// recursively by `Packer` in strip-local coordinates (kerf only, no trims —
/// the trims were already consumed building `usable_w`/`usable_h`) and then
/// translated into bin coordinates.
#[allow(clippy::too_many_arguments)]
pub fn strip_fill(
    pool: &[Piece],
    spec_width: f64,
    spec_height: f64,
    params: &CuttingParams,
    horizontal: bool,
    first_dim: Option<f64>,
    max_repeat: Option<u32>,
    min_rect_size: f64,
) -> Option<Fill> {
    let kerf = params.kerf.max(0.0);
    let left = params.left_trim.max(0.0);
    let right = params.right_trim.max(0.0);
    let top = params.top_trim.max(0.0);
    let bottom = params.bottom_trim.max(0.0);
    let usable_w = spec_width - left - right;
    let usable_h = spec_height - top - bottom;
    if usable_w <= 0.0 || usable_h <= 0.0 {
        return None;
    }

    // Strips advance along `axis_room` and span `strip_span`.
    let axis_room = if horizontal { usable_h } else { usable_w };
    let strip_span = if horizontal { usable_w } else { usable_h };

    let strip_params = CuttingParams {
        kerf,
        top_trim: 0.0,
        bottom_trim: 0.0,
        left_trim: 0.0,
        right_trim: 0.0,
    };

    let mut remaining: Vec<Piece> = pool.to_vec();
    let mut all_placed: Vec<PlacedPiece> = Vec::new();
    let mut all_cuts: Vec<Cut> = Vec::new();
    let mut all_rects: Vec<Rect> = Vec::new();
    let mut cursor = 0.0_f64;
    let mut room = axis_room;
    let mut first = true;
    let mut repeat_counts: HashMap<TypeSignature, u32> = HashMap::new();

    // Reused across strip candidates so the inner loop allocates once.
    let mut fitting: Vec<Piece> = Vec::with_capacity(pool.len());

    while !remaining.is_empty() && room >= min_rect_size {
        let mut candidates = strip_candidate_dims(&remaining, room, horizontal);
        if candidates.is_empty() {
            break;
        }
        if !candidates.iter().any(|d| *d == room) {
            candidates.push(room);
        }
        if first {
            if let Some(fd) = first_dim {
                if fd <= room {
                    candidates = vec![fd];
                }
                // Cleared whether or not the seed fitted — as in Python, where
                // `first = False` sits outside the `first_dim <= room` test.
                first = false;
            }
        }

        let mut best: Option<Strip> = None;

        for &size in &candidates {
            let (mat_w, mat_h) = if horizontal {
                (strip_span, size)
            } else {
                (size, strip_span)
            };
            let mut packer = match Packer::new(
                mat_w,
                mat_h,
                &strip_params,
                SplitRule::ShorterLeftoverAxis,
                Selection::MaxEfficiency,
                min_rect_size,
            ) {
                Some(p) => p,
                None => continue,
            };

            fitting.clear();
            fitting.extend(remaining.iter().copied().filter(|p| {
                (p.width <= mat_w && p.height <= mat_h)
                    || (p.can_rotate && p.height <= mat_w && p.width <= mat_h)
            }));
            if fitting.is_empty() {
                continue;
            }

            // Widest-fit-first inside the strip (rotation-aware); the nested
            // packer still nests narrower sub-columns recursively.
            let (limit, across) = if horizontal {
                (mat_h, true)
            } else {
                (mat_w, false)
            };
            fitting.sort_by(|a, b| {
                b.priority
                    .cmp(&a.priority)
                    .then_with(|| {
                        fit_across(b, limit, across)
                            .partial_cmp(&fit_across(a, limit, across))
                            .unwrap()
                    })
                    .then_with(|| {
                        let (sa, sb) = if horizontal {
                            (a.width, b.width)
                        } else {
                            (a.height, b.height)
                        };
                        sb.partial_cmp(&sa).unwrap()
                    })
                    .then_with(|| a.id_rank.cmp(&b.id_rank))
            });

            packer.optimize(&fitting);
            if packer.placed.is_empty() {
                continue;
            }

            let signature = match max_repeat {
                Some(cap) => {
                    let sig = type_signature(&packer.placed, pool);
                    if repeat_counts.get(&sig).copied().unwrap_or(0) >= cap {
                        continue;
                    }
                    Some(sig)
                }
                None => None,
            };

            let used: f64 = packer.placed.iter().map(|p| p.width * p.height).sum();
            let score = (used / (mat_w * mat_h), used, -size);
            if best.is_none() || score_gt(score, best.as_ref().unwrap().score) {
                best = Some(Strip {
                    score,
                    size,
                    placed: packer.placed,
                    cuts: packer.cuts,
                    remainders: packer.remainders,
                    signature,
                });
            }
        }

        let best = match best {
            Some(b) => b,
            None => break,
        };

        if let Some(sig) = best.signature {
            *repeat_counts.entry(sig).or_insert(0) += 1;
        }
        let size = best.size;
        let (dx, dy) = if horizontal {
            (left, bottom + cursor)
        } else {
            (left + cursor, bottom)
        };

        // `_translate`: strip-local geometry into bin coordinates.
        all_placed.extend(best.placed.iter().map(|pp| PlacedPiece {
            index: pp.index,
            x: pp.x + dx,
            y: pp.y + dy,
            width: pp.width,
            height: pp.height,
            rotated: pp.rotated,
        }));
        all_cuts.extend(best.cuts.iter().map(|c| Cut {
            x: c.x + dx,
            y: c.y + dy,
            length: c.length,
            is_horizontal: c.is_horizontal,
        }));
        all_rects.extend(
            best.remainders
                .iter()
                .map(|r| Rect::new(r.x + dx, r.y + dy, r.width, r.height)),
        );

        let leftover = room - size;
        let advance = if leftover > 0.0 {
            // First-stage cut separating this strip from the rest of the bin.
            if horizontal {
                all_cuts.push(Cut {
                    x: left,
                    y: bottom + cursor + size,
                    length: usable_w,
                    is_horizontal: true,
                });
            } else {
                all_cuts.push(Cut {
                    x: left + cursor + size,
                    y: bottom,
                    length: usable_h,
                    is_horizontal: false,
                });
            }
            size + kerf
        } else {
            size
        };
        cursor += advance;
        room -= advance;

        let taken: HashSet<u32> = best.placed.iter().map(|pp| pp.index).collect();
        remaining.retain(|p| !taken.contains(&p.index));
    }

    if all_placed.is_empty() {
        return None;
    }

    if room >= min_rect_size {
        // Unopened tail of the bin: one continuous first-stage leftover.
        if horizontal {
            all_rects.push(Rect::new(left, bottom + cursor, usable_w, room));
        } else {
            all_rects.push(Rect::new(left + cursor, bottom, room, usable_h));
        }
    }

    Some(Fill {
        placed: all_placed,
        remainders: all_rects,
        cuts: all_cuts,
    })
}

// ---------------------------------------------------------------------------
// gen_fills
// ---------------------------------------------------------------------------

/// Port of `search._Searcher._strip_seeds`.
///
/// The trims are read raw here (not clamped at 0) because that is what the
/// Python does — `strip_fill` is the one that clamps them.
fn strip_seeds(pool: &[Piece], spec_width: f64, params: &CuttingParams) -> Vec<f64> {
    let mut dims: Vec<f64> = Vec::with_capacity(pool.len() * 2);
    for p in pool {
        dims.push(p.width);
        if p.can_rotate {
            dims.push(p.height);
        }
    }
    dims.sort_by(|a, b| b.partial_cmp(a).unwrap());
    dims.dedup();
    let usable_w = spec_width - params.left_trim - params.right_trim;
    dims.into_iter().filter(|d| *d <= usable_w).take(4).collect()
}

/// One point of the greedy portfolio, in the exploration order the caller set
/// (the search reshuffles it under a seed to surface alternative solutions).
pub type PortfolioEntry = (SortKey, SplitRule, Selection);

/// Port of `search._Searcher.gen_fills`: every candidate fill of one bin from
/// one pool, deduped by placed-piece-type multiset.
///
/// This is where the FFI boundary belongs. Crossing it was measured at 32% of
/// the Rust-side time when the bridge called `greedy_fill` once per config; one
/// call per `(pool, spec)` returning up to `tries` fills pays that toll ~60x
/// less often.
pub fn gen_fills(
    pool: &[Piece],
    spec_width: f64,
    spec_height: f64,
    params: &CuttingParams,
    portfolio: &[PortfolioEntry],
    tries: u32,
    min_rect_size: f64,
) -> Vec<Fill> {
    let mut fills: Vec<Fill> = Vec::new();
    let mut seen: HashSet<TypeSignature> = HashSet::new();
    let mut spent: u32 = 0;

    let mut push = |fills: &mut Vec<Fill>, fill: Option<Fill>| {
        let fill = match fill {
            Some(f) if !f.placed.is_empty() => f,
            _ => return,
        };
        if seen.insert(type_signature(&fill.placed, pool)) {
            fills.push(fill);
        }
    };

    let strip_budget = 6.max(tries / 2);
    let seeds = strip_seeds(pool, spec_width, params);

    for horizontal in [false, true] {
        for first_dim in std::iter::once(None).chain(seeds.iter().map(|d| Some(*d))) {
            if spent >= strip_budget {
                break;
            }
            push(
                &mut fills,
                strip_fill(
                    pool,
                    spec_width,
                    spec_height,
                    params,
                    horizontal,
                    first_dim,
                    None,
                    min_rect_size,
                ),
            );
            spent += 1;
        }
    }

    // Repeat-capped variants: spread near-perfect single-type strips across
    // boards instead of hoarding them into one (see `strip_fill`).
    for horizontal in [false, true] {
        for repeat_cap in [1u32, 2u32] {
            if spent >= strip_budget + 4 {
                break;
            }
            push(
                &mut fills,
                strip_fill(
                    pool,
                    spec_width,
                    spec_height,
                    params,
                    horizontal,
                    None,
                    Some(repeat_cap),
                    min_rect_size,
                ),
            );
            spent += 1;
        }
    }

    for &(sort_key, split_rule, selection) in portfolio {
        if spent >= tries {
            break;
        }
        push(
            &mut fills,
            greedy_fill(
                pool,
                spec_width,
                spec_height,
                params,
                sort_key,
                split_rule,
                selection,
                min_rect_size,
            ),
        );
        spent += 1;
    }

    fills
}

/// Port of the probe loop inside `search._Searcher.complete_probe`: the first
/// constructor that places the WHOLE pool (`target` pieces) into one bin.
///
/// The order is the contract — the caller's first eight greedy configs, then
/// the two strip orientations. Python builds all ten candidates and then scans
/// for the first complete one; stopping at the first hit is equivalent (the
/// constructors share no state) and is most of the point of collapsing ten FFI
/// crossings into one.
pub fn probe_fill(
    pool: &[Piece],
    spec_width: f64,
    spec_height: f64,
    params: &CuttingParams,
    portfolio: &[PortfolioEntry],
    target: usize,
    min_rect_size: f64,
) -> Option<Fill> {
    for &(sort_key, split_rule, selection) in portfolio {
        let fill = greedy_fill(
            pool,
            spec_width,
            spec_height,
            params,
            sort_key,
            split_rule,
            selection,
            min_rect_size,
        );
        if let Some(fill) = fill {
            if fill.placed.len() == target {
                return Some(fill);
            }
        }
    }
    for horizontal in [false, true] {
        let fill = strip_fill(
            pool,
            spec_width,
            spec_height,
            params,
            horizontal,
            None,
            None,
            min_rect_size,
        );
        if let Some(fill) = fill {
            if fill.placed.len() == target {
                return Some(fill);
            }
        }
    }
    None
}
